### PR TITLE
Rwa 3103 Reconfiguration fixes

### PR DESCRIPTION
**Pull request by skandala-cgi**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/RWA-3119

### Change description ###

Updated reconfiguration for copying additional properties to top level properties, allowing null values to be set, for additional properties check canReconfigure flag before updating

**Does this PR introduce a breaking change?** (check one with "x")

```
[] Yes
[x] No
```

### CODE REVIEW

File: src/test/java/uk/gov/hmcts/reform/wataskmanagementapi/services/CaseConfigurationProviderServiceTest.java [hunk 1]
```diff
@@ -1075,6 +1077,136 @@ void should_not_recalculate_due_date_interval_from_given_due_date_origin_propert
...
+                                stringValue("additionalProperties_name4"),
+                                stringValue("value4")
+                        ),
+                        new ConfigurationDmnEvaluationResponse(
+                                stringValue("additionalProperties_name5"),
+                                stringValue("value5")
+                        ),
+                        new ConfigurationDmnEvaluationResponse(
+                            stringValue("additionalProperties_name6"), stringValue("value6"))
+                ));
```
Comment (raghera) [inline src/test/java/uk/gov/hmcts/reform/wataskmanagementapi/services/CaseConfigurationProviderServiceTest.java:1133]: So if you're not setting the canReconfigure here I assume the default is `null` or false?  So I would expect none of them to be reconfigured unless I am missing something?

File: src/test/java/uk/gov/hmcts/reform/wataskmanagementapi/services/CFTTaskMapperTest.java [hunk 2]
```diff
@@ -2418,6 +2413,36 @@ void should_map_task_role_permissions_for_granular_permissions() {
...
+                        booleanValue(true)
+                )
+        );
+
+        cftTaskMapper.reconfigureAdditionalTaskAttribute(taskResource, additionalPropertyDMNResponse);
+
+        assertEquals("roleAssignmentId", taskResource.getAdditionalProperties().get("roleAssignmentId"));
+        assertEquals("", taskResource.getAdditionalProperties().get("key3"));
+        assertEquals(null, taskResource.getAdditionalProperties().get("key4"));
+        assertEquals(3, taskResource.getAdditionalProperties().size());
```
[the remark this answers is not in the document]
Comment (raghera) [reply src/test/java/uk/gov/hmcts/reform/wataskmanagementapi/services/CFTTaskMapperTest.java:2442]: I think from what you showed me this is covered elsewhere however it probably means we have some refactoring to do in the code which is outside the scope of this PR.

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/CFTTaskMapper.java [hunk 3]
```diff
@@ -642,6 +652,36 @@ protected void reconfigureTaskAttribute(TaskResource taskResource,
...
+        }
+
+        configurationAdditionalAttributeDmnResponse.stream().filter(response -> response.getCanReconfigure() != null
+                && response.getCanReconfigure().getValue()).forEach(
+                    response -> {
+                        if (response.getName().getValue() != null && response.getValue() != null) {
+                            additionalProperties.put(response.getName().getValue()
+                                    .replace(ADDITIONAL_PROPERTIES_PREFIX, ""),
+                                    Optional.ofNullable(response.getValue().getValue()));
+                        }
```
[the remark this answers is not in the document]
Comment (raghera) [reply src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/CFTTaskMapper.java:674]: We have decided to adhere what the DMN provides.  An empty string is maintained and stored as such in the DB.  This will also be the returned in the response in json if it is a field that's is returned.

File: src/functionalTest/java/uk/gov/hmcts/reform/wataskmanagementapi/watasks/controllers/PostTaskExecuteReconfigureControllerTest.java [hunk 4]
```diff
@@ -480,6 +483,368 @@ public void should_recalculate_next_hearing_date_using_interval_calculation_when
...
+        );
+
+
+        result.prettyPrint();
+
+        result.then().assertThat()
+                .statusCode(HttpStatus.OK.value())
+                .and().contentType(MediaType.APPLICATION_JSON_VALUE)
+                .and().body("task.id", equalTo(taskId))
+                // .body("task.task_state", is("assigned"))
```
Comment (BenNzewi) [inline src/functionalTest/java/uk/gov/hmcts/reform/wataskmanagementapi/watasks/controllers/PostTaskExecuteReconfigureControllerTest.java:662]: Commented line? please introduce or remove

File: src/functionalTest/java/uk/gov/hmcts/reform/wataskmanagementapi/watasks/controllers/PostTaskExecuteReconfigureControllerTest.java [hunk 5]
```diff
@@ -480,6 +483,368 @@ public void should_recalculate_next_hearing_date_using_interval_calculation_when
...
+                assigneeCredentials.getHeaders()
+        );
+
+        result.prettyPrint();
+
+        result.then().assertThat()
+                .statusCode(HttpStatus.OK.value())
+                .and().contentType(MediaType.APPLICATION_JSON_VALUE)
+                .and().body("task.id", equalTo(taskId))
+                // .body("task.task_state", is("assigned"))
```
Comment (BenNzewi) [inline src/functionalTest/java/uk/gov/hmcts/reform/wataskmanagementapi/watasks/controllers/PostTaskExecuteReconfigureControllerTest.java:693]: Commented line? please introduce or remove

File: src/functionalTest/java/uk/gov/hmcts/reform/wataskmanagementapi/watasks/controllers/PostTaskExecuteReconfigureControllerTest.java [hunk 6]
```diff
@@ -480,6 +483,368 @@ public void should_recalculate_next_hearing_date_using_interval_calculation_when
...
+        );
+
+
+        result.prettyPrint();
+
+        result.then().assertThat()
+                .statusCode(HttpStatus.OK.value())
+                .and().contentType(MediaType.APPLICATION_JSON_VALUE)
+                .and().body("task.id", equalTo(taskId))
+                // .body("task.task_state", is("assigned"))
```
Comment (BenNzewi) [inline src/functionalTest/java/uk/gov/hmcts/reform/wataskmanagementapi/watasks/controllers/PostTaskExecuteReconfigureControllerTest.java:772]: Commented line? please introduce or remove

File: src/functionalTest/java/uk/gov/hmcts/reform/wataskmanagementapi/watasks/controllers/PostTaskExecuteReconfigureControllerTest.java [hunk 7]
```diff
@@ -480,6 +483,368 @@ public void should_recalculate_next_hearing_date_using_interval_calculation_when
...
+                assigneeCredentials.getHeaders()
+        );
+
+        result.prettyPrint();
+
+        result.then().assertThat()
+                .statusCode(HttpStatus.OK.value())
+                .and().contentType(MediaType.APPLICATION_JSON_VALUE)
+                .and().body("task.id", equalTo(taskId))
+                // .body("task.task_state", is("assigned"))
```
Comment (BenNzewi) [inline src/functionalTest/java/uk/gov/hmcts/reform/wataskmanagementapi/watasks/controllers/PostTaskExecuteReconfigureControllerTest.java:803]: Commented line? please introduce or remove

File: src/test/java/uk/gov/hmcts/reform/wataskmanagementapi/services/CFTTaskMapperTest.java [hunk 8]
```diff
@@ -2418,6 +2413,36 @@ void should_map_task_role_permissions_for_granular_permissions() {
...
+                ),
+                new ConfigurationDmnEvaluationResponse(stringValue("additionalProperties_key2"), stringValue("value2"),
+                        null
+                ),
+                new ConfigurationDmnEvaluationResponse(stringValue("additionalProperties_key3"), stringValue(""),
+                        booleanValue(true)
+                ),
+                new ConfigurationDmnEvaluationResponse(stringValue("additionalProperties_key4"), stringValue(null),
+                        booleanValue(true)
+                )
```
Comment (BenNzewi) [inline src/test/java/uk/gov/hmcts/reform/wataskmanagementapi/services/CFTTaskMapperTest.java:2434]: would adding 2 additional edge cases when - canReconfigure is false and value is empty or null help show we have all edge cases covered?

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/CaseConfigurationProviderService.java [hunk 9]
```diff
@@ -65,16 +65,8 @@ public TaskConfigurationResults getCaseRelatedConfiguration(
...
-
-        Map<String,String> additionalPropertiesMap =
-            (Map<String, String>) taskAttributes.get(ADDITIONAL_PROPERTIES_KEY);
-        // Copy all additional properties as top level elements in task attributes
-        if (additionalPropertiesMap != null && !additionalPropertiesMap.isEmpty()) {
-            taskAttributes.putAll(additionalPropertiesMap);
-        }
-
         String taskAttributesString = writeValueAsString(taskAttributes);
+        log.debug("Case Configuration : case data {}", caseDataString);
```
Comment (raghera) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/CaseConfigurationProviderService.java:69]: Should have asked last time but ... could the Case data string have anything sensitive in there?

**Review by raghera — APPROVED** [review]

**Review by raghera — APPROVED** [review]
